From d8ccbd750900c88dc3bef1a94d672736ddb4ee35 Mon Sep 17 00:00:00 2001
From: Boris Brezillon <boris.brezillon@collabora.com>
Date: Thu, 23 Apr 2026 09:06:10 +0200
Subject: [PATCH 1/6] panfrost: Get rid of unused PAN_BO_ACCESS_xx flags

PAN_BO_ACCESS_{PRIVATE,SHARED} are neither set nor checked, so that one
is easy.

PAN_BO_ACCESS_{VERTEX_TILER,FRAGMENT} are set but never checked except
in panfrost_bo_wait() where we check if gpu_access is != 0, but because
those flags are never set without a R/W qualifier, the != 0 check still
works. Drop the flags along with the logic that was setting those flags,
and drop the stage parameter that was passed in multiple places to do
this selection.

Once this is done, we can move PAN_BO_ACCESS_{READ,WRITE} to bit 0 and 1.

Signed-off-by: Boris Brezillon <boris.brezillon@collabora.com>
Reviewed-by: Eric Guo <Eric.guo@nxp.com>
---
 src/gallium/drivers/panfrost/pan_bo.h        | 15 +----
 src/gallium/drivers/panfrost/pan_cmdstream.c | 65 ++++++++++----------
 src/gallium/drivers/panfrost/pan_context.h   |  1 -
 src/gallium/drivers/panfrost/pan_csf.c       |  6 +-
 src/gallium/drivers/panfrost/pan_helpers.c   |  7 +--
 src/gallium/drivers/panfrost/pan_jm.c        |  4 +-
 src/gallium/drivers/panfrost/pan_job.c       | 35 ++++-------
 src/gallium/drivers/panfrost/pan_job.h       | 12 ++--
 8 files changed, 56 insertions(+), 89 deletions(-)

diff --git a/src/gallium/drivers/panfrost/pan_bo.h b/src/gallium/drivers/panfrost/pan_bo.h
index 0372c05cbc30..b3a5d9f58983 100644
--- a/src/gallium/drivers/panfrost/pan_bo.h
+++ b/src/gallium/drivers/panfrost/pan_bo.h
@@ -40,22 +40,11 @@
 
 /* GPU access flags */
 
-/* BO is either shared (can be accessed by more than one GPU batch) or private
- * (reserved by a specific GPU job). */
-#define PAN_BO_ACCESS_PRIVATE (0 << 0)
-#define PAN_BO_ACCESS_SHARED  (1 << 0)
-
 /* BO is being read/written by the GPU */
-#define PAN_BO_ACCESS_READ  (1 << 1)
-#define PAN_BO_ACCESS_WRITE (1 << 2)
+#define PAN_BO_ACCESS_READ  (1 << 0)
+#define PAN_BO_ACCESS_WRITE (1 << 1)
 #define PAN_BO_ACCESS_RW    (PAN_BO_ACCESS_READ | PAN_BO_ACCESS_WRITE)
 
-/* BO is accessed by the vertex/tiler job. */
-#define PAN_BO_ACCESS_VERTEX_TILER (1 << 3)
-
-/* BO is accessed by the fragment job. */
-#define PAN_BO_ACCESS_FRAGMENT (1 << 4)
-
 typedef uint8_t pan_bo_access;
 
 struct panfrost_device;
diff --git a/src/gallium/drivers/panfrost/pan_cmdstream.c b/src/gallium/drivers/panfrost/pan_cmdstream.c
index 6629934fecdf..8e8cd3f07d35 100644
--- a/src/gallium/drivers/panfrost/pan_cmdstream.c
+++ b/src/gallium/drivers/panfrost/pan_cmdstream.c
@@ -434,8 +434,8 @@ panfrost_emit_compute_shader_meta(struct panfrost_batch *batch,
 {
    struct panfrost_compiled_shader *ss = batch->ctx->prog[stage];
 
-   panfrost_batch_add_bo(batch, ss->bin.bo, MESA_SHADER_VERTEX);
-   panfrost_batch_add_bo(batch, ss->state.bo, MESA_SHADER_VERTEX);
+   panfrost_batch_add_bo(batch, ss->bin.bo);
+   panfrost_batch_add_bo(batch, ss->state.bo);
 
    return ss->state.gpu;
 }
@@ -674,8 +674,8 @@ panfrost_emit_frag_shader_meta(struct panfrost_batch *batch)
    struct panfrost_context *ctx = batch->ctx;
    struct panfrost_compiled_shader *ss = ctx->prog[MESA_SHADER_FRAGMENT];
 
-   panfrost_batch_add_bo(batch, ss->bin.bo, MESA_SHADER_FRAGMENT);
-   panfrost_batch_add_bo(batch, ss->state.bo, MESA_SHADER_FRAGMENT);
+   panfrost_batch_add_bo(batch, ss->bin.bo);
+   panfrost_batch_add_bo(batch, ss->state.bo);
 
    struct pan_ptr xfer;
 
@@ -959,7 +959,7 @@ panfrost_emit_vertex_buffers(struct panfrost_batch *batch)
       struct panfrost_resource *rsrc = pan_resource(prsrc);
       assert(!vb.is_user_buffer);
 
-      panfrost_batch_read_rsrc(batch, rsrc, MESA_SHADER_VERTEX);
+      panfrost_batch_read_rsrc(batch, rsrc);
 
       pan_pack(buffers + i, BUFFER, cfg) {
          cfg.address = rsrc->plane.base + vb.buffer_offset;
@@ -1018,7 +1018,7 @@ panfrost_emit_images(struct panfrost_batch *batch, mesa_shader_stage stage)
       panfrost_update_sampler_view(&view, &ctx->base);
       out[i] = view.bifrost_tex_descriptor;
 
-      panfrost_track_image_access(batch, stage, image);
+      panfrost_track_image_access(batch, image);
    }
 
    return T.gpu;
@@ -1027,7 +1027,6 @@ panfrost_emit_images(struct panfrost_batch *batch, mesa_shader_stage stage)
 
 static uint64_t
 panfrost_map_constant_buffer_gpu(struct panfrost_batch *batch,
-                                 mesa_shader_stage st,
                                  struct panfrost_constant_buffer *buf,
                                  unsigned index)
 {
@@ -1035,7 +1034,7 @@ panfrost_map_constant_buffer_gpu(struct panfrost_batch *batch,
    struct panfrost_resource *rsrc = pan_resource(cb->buffer);
 
    if (rsrc) {
-      panfrost_batch_read_rsrc(batch, rsrc, st);
+      panfrost_batch_read_rsrc(batch, rsrc);
 
       /* Alignment gauranteed by
        * pipe_caps.constant_buffer_offset_alignment */
@@ -1191,7 +1190,7 @@ panfrost_upload_ssbo_sysval(struct panfrost_batch *batch,
    struct panfrost_resource *rsrc = pan_resource(sb.buffer);
    struct panfrost_bo *bo = rsrc->bo;
 
-   panfrost_batch_write_rsrc(batch, rsrc, st);
+   panfrost_batch_write_rsrc(batch, rsrc);
 
    util_range_add(&rsrc->base, &rsrc->valid_buffer_range, sb.buffer_offset,
                   sb.buffer_size);
@@ -1363,7 +1362,7 @@ panfrost_upload_sysvals(struct panfrost_batch *batch, void *ptr_cpu,
          util_range_add(&rsrc->base, &rsrc->valid_buffer_range, offset,
                         target->buffer_size - offset);
 
-         panfrost_batch_write_rsrc(batch, rsrc, MESA_SHADER_VERTEX);
+         panfrost_batch_write_rsrc(batch, rsrc);
 
          uniforms[i].du[0] = rsrc->plane.base + offset;
          break;
@@ -1480,7 +1479,7 @@ panfrost_emit_ssbos(struct panfrost_batch *batch, mesa_shader_stage st)
       struct panfrost_resource *rsrc = pan_resource(sb.buffer);
       struct panfrost_bo *bo = rsrc->bo;
 
-      panfrost_batch_write_rsrc(batch, rsrc, st);
+      panfrost_batch_write_rsrc(batch, rsrc);
 
       util_range_add(&rsrc->base, &rsrc->valid_buffer_range, sb.buffer_offset,
                      sb.buffer_size);
@@ -1572,7 +1571,7 @@ panfrost_emit_const_buf(struct panfrost_batch *batch,
 
       if (usz > 0) {
          address =
-            panfrost_map_constant_buffer_gpu(batch, stage, buf, ubo_adj);
+            panfrost_map_constant_buffer_gpu(batch, buf, ubo_adj);
       }
 
       panfrost_emit_ubo(ubos.cpu, ubo, address, usz);
@@ -1702,7 +1701,7 @@ panfrost_emit_shared_memory(struct panfrost_batch *batch,
 
 #if PAN_ARCH <= 5
 static uint64_t
-panfrost_get_tex_desc(struct panfrost_batch *batch, mesa_shader_stage st,
+panfrost_get_tex_desc(struct panfrost_batch *batch,
                       struct panfrost_sampler_view *view)
 {
    if (!view)
@@ -1711,8 +1710,8 @@ panfrost_get_tex_desc(struct panfrost_batch *batch, mesa_shader_stage st,
    struct pipe_sampler_view *pview = &view->base;
    struct panfrost_resource *rsrc = pan_resource(pview->texture);
 
-   panfrost_batch_read_rsrc(batch, rsrc, st);
-   panfrost_batch_add_bo(batch, view->state.bo, st);
+   panfrost_batch_read_rsrc(batch, rsrc);
+   panfrost_batch_add_bo(batch, view->state.bo);
 
    return view->state.gpu;
 }
@@ -1978,8 +1977,8 @@ panfrost_emit_texture_descriptors(struct panfrost_batch *batch,
       out[i] = view->bifrost_tex_descriptor;
 #endif
 
-      panfrost_batch_read_rsrc(batch, rsrc, stage);
-      panfrost_batch_add_bo(batch, view->state.bo, stage);
+      panfrost_batch_read_rsrc(batch, rsrc);
+      panfrost_batch_add_bo(batch, view->state.bo);
    }
 
    for (int i = actual_count; i < needed_count; ++i)
@@ -1999,7 +1998,7 @@ panfrost_emit_texture_descriptors(struct panfrost_batch *batch,
 
       panfrost_update_sampler_view(view, &ctx->base);
 
-      trampolines[i] = panfrost_get_tex_desc(batch, stage, view);
+      trampolines[i] = panfrost_get_tex_desc(batch, view);
    }
 
    for (int i = actual_count; i < needed_count; ++i)
@@ -2153,7 +2152,7 @@ emit_image_bufs(struct panfrost_batch *batch, mesa_shader_stage shader,
                                      : layout->array_stride_B));
       }
 
-      panfrost_track_image_access(batch, shader, image);
+      panfrost_track_image_access(batch, image);
 
 #if PAN_ARCH >= 6
       if (is_buffer) {
@@ -2383,7 +2382,7 @@ panfrost_emit_vertex_data(struct panfrost_batch *batch, uint64_t *buffers)
       if (!rsrc)
          continue;
 
-      panfrost_batch_read_rsrc(batch, rsrc, MESA_SHADER_VERTEX);
+      panfrost_batch_read_rsrc(batch, rsrc);
 
       /* Mask off lower bits, see offset fixup below */
       uint64_t raw_addr = rsrc->plane.base + buf->buffer_offset;
@@ -3587,8 +3586,7 @@ panfrost_draw_indirect(struct pipe_context *pipe,
 
    struct pipe_draw_info tmp_info = *info;
 
-   panfrost_batch_read_rsrc(batch, pan_resource(indirect->buffer),
-                            MESA_SHADER_VERTEX);
+   panfrost_batch_read_rsrc(batch, pan_resource(indirect->buffer));
 
    panfrost_update_active_prim(ctx, info->mode);
 
@@ -3598,7 +3596,7 @@ panfrost_draw_indirect(struct pipe_context *pipe,
    if (info->index_size) {
       struct panfrost_resource *index_buffer =
          pan_resource(info->index.resource);
-      panfrost_batch_read_rsrc(batch, index_buffer, MESA_SHADER_VERTEX);
+      panfrost_batch_read_rsrc(batch, index_buffer);
       batch->indices = index_buffer->plane.base;
    }
 
@@ -3746,7 +3744,7 @@ panfrost_launch_grid_on_batch(struct pipe_context *pipe,
          continue;
 
       struct panfrost_resource *buffer = pan_resource(*res);
-      panfrost_batch_write_rsrc(batch, buffer, MESA_SHADER_COMPUTE);
+      panfrost_batch_write_rsrc(batch, buffer);
    }
 
    if (info->indirect && !PAN_GPU_SUPPORTS_DISPATCH_INDIRECT) {
@@ -3784,7 +3782,7 @@ panfrost_launch_grid_on_batch(struct pipe_context *pipe,
 
    /* if indirect, mark the indirect buffer as being read */
    if (info->indirect)
-      panfrost_batch_read_rsrc(batch, pan_resource(info->indirect), MESA_SHADER_COMPUTE);
+      panfrost_batch_read_rsrc(batch, pan_resource(info->indirect));
 
    /* launch it */
 #if PAN_ARCH >= 10
@@ -3902,8 +3900,8 @@ panfrost_afbc_size(struct panfrost_batch *batch, struct panfrost_resource *src,
                      src->image.props.modifier,
                      u_minify(src->image.props.extent_px.height, level));
 
-   panfrost_batch_read_rsrc(batch, src, MESA_SHADER_COMPUTE);
-   panfrost_batch_write_bo(batch, layout, MESA_SHADER_COMPUTE);
+   panfrost_batch_read_rsrc(batch, src);
+   panfrost_batch_write_bo(batch, layout);
 
    LAUNCH_AFBC_CONV_SHADER(size, batch, src, consts, nr_sblocks);
 }
@@ -3938,9 +3936,9 @@ panfrost_afbc_pack(struct panfrost_batch *batch, struct panfrost_resource *src,
       .dst_stride = dst_stride_sb,
    };
 
-   panfrost_batch_read_rsrc(batch, src, MESA_SHADER_COMPUTE);
-   panfrost_batch_write_bo(batch, dst, MESA_SHADER_COMPUTE);
-   panfrost_batch_add_bo(batch, layout, MESA_SHADER_COMPUTE);
+   panfrost_batch_read_rsrc(batch, src);
+   panfrost_batch_write_bo(batch, dst);
+   panfrost_batch_add_bo(batch, layout);
 
    LAUNCH_AFBC_CONV_SHADER(pack, batch, src, consts, nr_sblocks);
 }
@@ -3976,8 +3974,8 @@ panfrost_compute_copy_buffer(struct pipe_context *pctx,
 
    struct panfrost_batch *batch = panfrost_get_batch_for_fbo(ctx);
 
-   panfrost_batch_read_rsrc(batch, src, MESA_SHADER_COMPUTE);
-   panfrost_batch_write_rsrc(batch, dst, MESA_SHADER_COMPUTE);
+   panfrost_batch_read_rsrc(batch, src);
+   panfrost_batch_write_rsrc(batch, dst);
 
    /* PANLIB_BARRIER_JM_BARRIER only takes effect on the JM path (v6-v9);
     * panfrost_launch_precomp ignores the barrier argument on CSF (v10+).
@@ -4729,8 +4727,7 @@ batch_get_polygon_list(struct panfrost_batch *batch)
       }
 
       batch->tiler_ctx.midgard.polygon_list = batch->polygon_list_bo->ptr.gpu;
-      panfrost_batch_add_bo(batch, batch->polygon_list_bo,
-                            MESA_SHADER_FRAGMENT);
+      panfrost_batch_add_bo(batch, batch->polygon_list_bo);
 
       if (init_polygon_list && dev->model->quirks.no_hierarchical_tiling) {
          assert(batch->polygon_list_bo->ptr.cpu);
diff --git a/src/gallium/drivers/panfrost/pan_context.h b/src/gallium/drivers/panfrost/pan_context.h
index 75e72ab6c777..133c84ad8b8e 100644
--- a/src/gallium/drivers/panfrost/pan_context.h
+++ b/src/gallium/drivers/panfrost/pan_context.h
@@ -610,7 +610,6 @@ void panfrost_set_batch_masks_blend(struct panfrost_batch *batch);
 void panfrost_set_batch_masks_zs(struct panfrost_batch *batch);
 
 void panfrost_track_image_access(struct panfrost_batch *batch,
-                                 mesa_shader_stage stage,
                                  struct pipe_image_view *image);
 
 void panfrost_context_reinit(struct panfrost_context *ctx);
diff --git a/src/gallium/drivers/panfrost/pan_csf.c b/src/gallium/drivers/panfrost/pan_csf.c
index 3f830802f508..4b7607942b50 100644
--- a/src/gallium/drivers/panfrost/pan_csf.c
+++ b/src/gallium/drivers/panfrost/pan_csf.c
@@ -1614,7 +1614,7 @@ csf_emit_draw_state(struct panfrost_batch *batch,
    if (panfrost_occlusion_query_active(ctx)) {
       struct panfrost_resource *rsrc = pan_resource(ctx->occlusion_query->rsrc);
       cs_move64_to(b, cs_sr_reg64(b, IDVS, OQ), rsrc->plane.base);
-      panfrost_batch_write_rsrc(ctx->batch, rsrc, MESA_SHADER_FRAGMENT);
+      panfrost_batch_write_rsrc(ctx->batch, rsrc);
    }
 
    cs_move32_to(b, cs_sr_reg32(b, IDVS, VARY_SIZE),
@@ -2159,7 +2159,7 @@ GENX(csf_emit_write_timestamp)(struct panfrost_batch *batch,
       : cs_now();
    cs_store_state(b, address, 0, MALI_CS_STATE_TIMESTAMP, async);
 
-   panfrost_batch_write_rsrc(batch, dst, MESA_SHADER_VERTEX);
+   panfrost_batch_write_rsrc(batch, dst);
 }
 
 void
@@ -2189,5 +2189,5 @@ GENX(csf_emit_copy_data)(struct panfrost_batch *batch,
    cs_wait_slot(b, 0);
    cs_store(b, data, dst_addr, BITFIELD_MASK(count), 0);
 
-   panfrost_batch_write_rsrc(batch, dst, MESA_SHADER_VERTEX);
+   panfrost_batch_write_rsrc(batch, dst);
 }
diff --git a/src/gallium/drivers/panfrost/pan_helpers.c b/src/gallium/drivers/panfrost/pan_helpers.c
index 682775cf89f6..c4b0c16a7948 100644
--- a/src/gallium/drivers/panfrost/pan_helpers.c
+++ b/src/gallium/drivers/panfrost/pan_helpers.c
@@ -87,7 +87,7 @@ panfrost_get_index_buffer(struct panfrost_batch *batch,
 
    if (!info->has_user_indices) {
       /* Only resources can be directly mapped */
-      panfrost_batch_read_rsrc(batch, rsrc, MESA_SHADER_VERTEX);
+      panfrost_batch_read_rsrc(batch, rsrc);
       return rsrc->plane.base + offset;
    } else {
       /* Otherwise, we need to upload to transient memory */
@@ -261,13 +261,12 @@ panfrost_set_batch_masks_zs(struct panfrost_batch *batch)
 
 void
 panfrost_track_image_access(struct panfrost_batch *batch,
-                            mesa_shader_stage stage,
                             struct pipe_image_view *image)
 {
    struct panfrost_resource *rsrc = pan_resource(image->resource);
 
    if (image->shader_access & PIPE_IMAGE_ACCESS_WRITE) {
-      panfrost_batch_write_rsrc(batch, rsrc, stage);
+      panfrost_batch_write_rsrc(batch, rsrc);
 
       bool is_buffer = rsrc->base.target == PIPE_BUFFER;
       unsigned level = is_buffer ? 0 : image->u.tex.level;
@@ -278,6 +277,6 @@ panfrost_track_image_access(struct panfrost_batch *batch,
                         rsrc->base.width0);
       }
    } else {
-      panfrost_batch_read_rsrc(batch, rsrc, stage);
+      panfrost_batch_read_rsrc(batch, rsrc);
    }
 }
diff --git a/src/gallium/drivers/panfrost/pan_jm.c b/src/gallium/drivers/panfrost/pan_jm.c
index a257506e7904..6d2c0453f5c3 100644
--- a/src/gallium/drivers/panfrost/pan_jm.c
+++ b/src/gallium/drivers/panfrost/pan_jm.c
@@ -554,7 +554,7 @@ jm_emit_tiler_draw(struct mali_draw_packed *out, struct panfrost_batch *batch,
          struct panfrost_resource *rsrc =
             pan_resource(ctx->occlusion_query->rsrc);
          cfg.occlusion = rsrc->plane.base;
-         panfrost_batch_write_rsrc(ctx->batch, rsrc, MESA_SHADER_FRAGMENT);
+         panfrost_batch_write_rsrc(ctx->batch, rsrc);
       }
 
 #if PAN_ARCH >= 9
@@ -1104,7 +1104,7 @@ GENX(jm_emit_write_timestamp)(struct panfrost_batch *batch,
 
    pan_jc_add_job(&batch->jm.jobs.vtc_jc, MALI_JOB_TYPE_WRITE_VALUE, false,
                   false, 0, 0, &job, false);
-   panfrost_batch_write_rsrc(batch, dst, MESA_SHADER_VERTEX);
+   panfrost_batch_write_rsrc(batch, dst);
 }
 
 int
diff --git a/src/gallium/drivers/panfrost/pan_job.c b/src/gallium/drivers/panfrost/pan_job.c
index 6dfde4e2933d..26dbac5fe545 100644
--- a/src/gallium/drivers/panfrost/pan_job.c
+++ b/src/gallium/drivers/panfrost/pan_job.c
@@ -49,7 +49,7 @@ panfrost_batch_add_surface(struct panfrost_batch *batch,
       struct panfrost_resource *rsrc = pan_resource(surf->texture);
       pan_resource_modifier_legalize(batch->ctx, rsrc, surf->format, true,
                                      false);
-      panfrost_batch_write_rsrc(batch, rsrc, MESA_SHADER_FRAGMENT);
+      panfrost_batch_write_rsrc(batch, rsrc);
    }
 }
 
@@ -336,35 +336,23 @@ panfrost_batch_add_bo_old(struct panfrost_batch *batch, struct panfrost_bo *bo,
    *entry = flags;
 }
 
-static uint32_t
-panfrost_access_for_stage(mesa_shader_stage stage)
-{
-   return (stage == MESA_SHADER_FRAGMENT) ? PAN_BO_ACCESS_FRAGMENT
-                                          : PAN_BO_ACCESS_VERTEX_TILER;
-}
-
 void
-panfrost_batch_add_bo(struct panfrost_batch *batch, struct panfrost_bo *bo,
-                      mesa_shader_stage stage)
+panfrost_batch_add_bo(struct panfrost_batch *batch, struct panfrost_bo *bo)
 {
-   panfrost_batch_add_bo_old(
-      batch, bo, PAN_BO_ACCESS_READ | panfrost_access_for_stage(stage));
+   panfrost_batch_add_bo_old(batch, bo, PAN_BO_ACCESS_READ);
 }
 
 void
-panfrost_batch_write_bo(struct panfrost_batch *batch, struct panfrost_bo *bo,
-                        mesa_shader_stage stage)
+panfrost_batch_write_bo(struct panfrost_batch *batch, struct panfrost_bo *bo)
 {
-   panfrost_batch_add_bo_old(
-      batch, bo, PAN_BO_ACCESS_WRITE | panfrost_access_for_stage(stage));
+   panfrost_batch_add_bo_old(batch, bo, PAN_BO_ACCESS_WRITE);
 }
 
 void
 panfrost_batch_read_rsrc(struct panfrost_batch *batch,
-                         struct panfrost_resource *rsrc,
-                         mesa_shader_stage stage)
+                         struct panfrost_resource *rsrc)
 {
-   uint32_t access = PAN_BO_ACCESS_READ | panfrost_access_for_stage(stage);
+   uint32_t access = PAN_BO_ACCESS_READ;
 
    pan_resource_update_access(batch->ctx, rsrc, false);
 
@@ -380,10 +368,9 @@ panfrost_batch_read_rsrc(struct panfrost_batch *batch,
 
 void
 panfrost_batch_write_rsrc(struct panfrost_batch *batch,
-                          struct panfrost_resource *rsrc,
-                          mesa_shader_stage stage)
+                          struct panfrost_resource *rsrc)
 {
-   uint32_t access = PAN_BO_ACCESS_WRITE | panfrost_access_for_stage(stage);
+   uint32_t access = PAN_BO_ACCESS_WRITE;
 
    pan_resource_update_access(batch->ctx, rsrc, true);
 
@@ -407,7 +394,7 @@ panfrost_batch_create_bo(struct panfrost_batch *batch, size_t size,
    bo = panfrost_bo_create(pan_device(batch->ctx->base.screen), size,
                            create_flags, label);
    if (bo) {
-      panfrost_batch_add_bo(batch, bo, stage);
+      panfrost_batch_add_bo(batch, bo);
 
       /* panfrost_batch_add_bo() has retained a reference and
        * panfrost_bo_create() initialize the refcnt to 1, so let's
@@ -435,7 +422,7 @@ panfrost_batch_get_scratchpad(struct panfrost_batch *batch,
                                   MESA_SHADER_VERTEX, "Thread local storage");
 
       if (batch->scratchpad)
-         panfrost_batch_add_bo(batch, batch->scratchpad, MESA_SHADER_FRAGMENT);
+         panfrost_batch_add_bo(batch, batch->scratchpad);
    }
 
    return batch->scratchpad;
diff --git a/src/gallium/drivers/panfrost/pan_job.h b/src/gallium/drivers/panfrost/pan_job.h
index 4e47406ac119..81e47cf36a4d 100644
--- a/src/gallium/drivers/panfrost/pan_job.h
+++ b/src/gallium/drivers/panfrost/pan_job.h
@@ -192,20 +192,16 @@ struct panfrost_batch *
 panfrost_get_fresh_batch_for_fbo(struct panfrost_context *ctx,
                                  const char *reason);
 
-void panfrost_batch_add_bo(struct panfrost_batch *batch, struct panfrost_bo *bo,
-                           mesa_shader_stage stage);
+void panfrost_batch_add_bo(struct panfrost_batch *batch, struct panfrost_bo *bo);
 
 void panfrost_batch_write_bo(struct panfrost_batch *batch,
-                             struct panfrost_bo *bo,
-                             mesa_shader_stage stage);
+                             struct panfrost_bo *bo);
 
 void panfrost_batch_read_rsrc(struct panfrost_batch *batch,
-                              struct panfrost_resource *rsrc,
-                              mesa_shader_stage stage);
+                              struct panfrost_resource *rsrc);
 
 void panfrost_batch_write_rsrc(struct panfrost_batch *batch,
-                               struct panfrost_resource *rsrc,
-                               mesa_shader_stage stage);
+                               struct panfrost_resource *rsrc);
 
 bool panfrost_any_batch_reads_rsrc(struct panfrost_context *ctx,
                                    struct panfrost_resource *rsrc);
-- 
GitLab


From d329d584e3fb19826ca590fc6b43fef20c8c5114 Mon Sep 17 00:00:00 2001
From: Boris Brezillon <boris.brezillon@collabora.com>
Date: Thu, 23 Apr 2026 17:18:21 +0200
Subject: [PATCH 2/6] panfrost: Drop the unused dump_headerblock() helper

Not used, never was => drop it.

Signed-off-by: Boris Brezillon <boris.brezillon@collabora.com>
Reviewed-by: Eric Guo <Eric.guo@nxp.com>
---
 src/gallium/drivers/panfrost/pan_resource.c | 29 ---------------------
 1 file changed, 29 deletions(-)

diff --git a/src/gallium/drivers/panfrost/pan_resource.c b/src/gallium/drivers/panfrost/pan_resource.c
index 178b7b8a5624..4a402ef6be89 100644
--- a/src/gallium/drivers/panfrost/pan_resource.c
+++ b/src/gallium/drivers/panfrost/pan_resource.c
@@ -1373,35 +1373,6 @@ panfrost_load_tiled_images(struct panfrost_transfer *transfer,
 
 #if MESA_DEBUG
 
-static void
-dump_headerblock(struct panfrost_resource *rsrc, uint32_t idx)
-{
-   panfrost_bo_wait(rsrc->bo, INT64_MAX, false);
-
-   uint8_t *ptr = rsrc->bo->ptr.cpu;
-   struct pan_afbc_headerblock *header = (struct pan_afbc_headerblock *)
-      (ptr + (idx * AFBC_HEADER_BYTES_PER_TILE));
-   uint32_t *header_u32 = (uint32_t *)header;
-   uint32_t *body = (uint32_t *)(ptr + header->payload.offset);
-   struct pan_image_block_size block_sz =
-      pan_afbc_subblock_size(rsrc->modifier);
-   unsigned pixel_sz = util_format_get_blocksize(rsrc->base.format);
-   unsigned uncompressed_size = pixel_sz * block_sz.width * block_sz.height;
-   uint32_t size = pan_afbc_payload_size(7, *header, uncompressed_size);
-
-   fprintf(stderr, "  Header: %08x %08x %08x %08x (size: %u bytes)\n",
-           header_u32[0], header_u32[1], header_u32[2], header_u32[3], size);
-   if (size > 0) {
-      fprintf(stderr, "  Body:   %08x %08x %08x %08x\n", body[0], body[1],
-              body[2], body[3]);
-   } else {
-      fprintf(stderr, "  Color:  0x%02x%02x%02x%02x\n",
-              header->color.rgba8888.r, header->color.rgba8888.g,
-              header->color.rgba8888.b, header->color.rgba8888.a);
-   }
-   fprintf(stderr, "\n");
-}
-
 void
 pan_dump_resource(struct panfrost_context *ctx, struct panfrost_resource *rsc)
 {
-- 
GitLab


From d03c70c07e7c6539f51828ff2fed27003a605dd9 Mon Sep 17 00:00:00 2001
From: Boris Brezillon <boris.brezillon@collabora.com>
Date: Fri, 24 Apr 2026 10:56:26 +0200
Subject: [PATCH 3/6] panfrost: Record access to plane1+ in case of
 multi-planar resources

At the moment we only track accesses to multiplanar resources as accesses
on the first plane. This has implications if the plane is then passed
to some other HW which waits on readiness. In practice it's probably
okay because all planes will be used at the same time, so waiting on
the first plane should be enough, but better safe than sorry.

Signed-off-by: Boris Brezillon <boris.brezillon@collabora.com>
Reviewed-by: Eric Guo <Eric.guo@nxp.com>
---
 src/gallium/drivers/panfrost/pan_job.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/src/gallium/drivers/panfrost/pan_job.c b/src/gallium/drivers/panfrost/pan_job.c
index 26dbac5fe545..a3da7ddd86ad 100644
--- a/src/gallium/drivers/panfrost/pan_job.c
+++ b/src/gallium/drivers/panfrost/pan_job.c
@@ -352,6 +352,12 @@ void
 panfrost_batch_read_rsrc(struct panfrost_batch *batch,
                          struct panfrost_resource *rsrc)
 {
+   /* Recursive call to panfrost_batch_read_rsrc() to add all planes.
+    * The max recursion depth should be 3.
+    */
+   if (rsrc->base.next)
+      panfrost_batch_read_rsrc(batch, pan_resource(rsrc->base.next));
+
    uint32_t access = PAN_BO_ACCESS_READ;
 
    pan_resource_update_access(batch->ctx, rsrc, false);
@@ -370,6 +376,12 @@ void
 panfrost_batch_write_rsrc(struct panfrost_batch *batch,
                           struct panfrost_resource *rsrc)
 {
+   /* Recursive call to panfrost_batch_write_rsrc() to add all planes.
+    * The max recursion depth should be 3.
+    */
+   if (rsrc->base.next)
+      panfrost_batch_write_rsrc(batch, pan_resource(rsrc->base.next));
+
    uint32_t access = PAN_BO_ACCESS_WRITE;
 
    pan_resource_update_access(batch->ctx, rsrc, true);
-- 
GitLab


From 4f00d7ee75b64a048e8639c6b861fc1f95f36635 Mon Sep 17 00:00:00 2001
From: Boris Brezillon <boris.brezillon@collabora.com>
Date: Thu, 23 Apr 2026 09:17:16 +0200
Subject: [PATCH 4/6] panfrost: Add a panfrost_resource_wait() helper

Prepare things for per-ctx BO access tracking by defining a
panfrost_resource_wait() taking both the resource and the context
the wait is targeted on. This will allow us to optimize waits if the
BO backing the resource is deemed ready according to the context, while
solving the resource-shared-across-contexts race that exist with the
panfrost::gpu_access flags.

Right now this is just a simple wrapper around panfrost_bo_wait(),
and the ctx parameter is unused. This will be addressed in a follow-up.

Signed-off-by: Boris Brezillon <boris.brezillon@collabora.com>
Reviewed-by: Eric Guo <Eric.guo@nxp.com>
---
 src/gallium/drivers/panfrost/pan_context.c  | 11 +++++----
 src/gallium/drivers/panfrost/pan_resource.c | 26 ++++++++++++++-------
 src/gallium/drivers/panfrost/pan_resource.h |  7 ++++++
 3 files changed, 30 insertions(+), 14 deletions(-)

diff --git a/src/gallium/drivers/panfrost/pan_context.c b/src/gallium/drivers/panfrost/pan_context.c
index 047c09520350..b48c9720947f 100644
--- a/src/gallium/drivers/panfrost/pan_context.c
+++ b/src/gallium/drivers/panfrost/pan_context.c
@@ -551,10 +551,11 @@ static const void *
 panfrost_trace_get_data(struct u_trace_context *utctx, void *buffer,
                         uint64_t offset_B, uint32_t size_B)
 {
+   struct panfrost_context *ctx = pan_context((struct pipe_context *)utctx->pctx);
    struct panfrost_resource *rsrc =
       pan_resource((struct pipe_resource *)buffer);
 
-   panfrost_bo_wait(rsrc->bo, INT64_MAX, false);
+   panfrost_resource_wait(rsrc, ctx, INT64_MAX, false);
 
    if (panfrost_bo_mmap(rsrc->bo))
       return NULL;
@@ -574,7 +575,7 @@ panfrost_trace_read_ts(struct u_trace_context *utctx, void *timestamps,
    if (!dev->kmod.dev->props.timestamp_frequency)
       return U_TRACE_NO_TIMESTAMP;
 
-   panfrost_bo_wait(rsrc->bo, INT64_MAX, false);
+   panfrost_resource_wait(rsrc, ctx, INT64_MAX, false);
 
    if (panfrost_bo_mmap(rsrc->bo))
       return U_TRACE_NO_TIMESTAMP;
@@ -815,7 +816,7 @@ panfrost_get_query_result(struct pipe_context *pipe, struct pipe_query *q,
    case PIPE_QUERY_OCCLUSION_PREDICATE:
    case PIPE_QUERY_OCCLUSION_PREDICATE_CONSERVATIVE:
       panfrost_flush_writer(ctx, rsrc, "Occlusion query");
-      panfrost_bo_wait(rsrc->bo, INT64_MAX, false);
+      panfrost_resource_wait(rsrc, ctx, INT64_MAX, false);
 
       /* Read back the query results */
       uint64_t *result = (uint64_t *)rsrc->bo->ptr.cpu;
@@ -837,7 +838,7 @@ panfrost_get_query_result(struct pipe_context *pipe, struct pipe_query *q,
 
    case PIPE_QUERY_TIMESTAMP: {
       panfrost_flush_writer(ctx, rsrc, "Timestamp query");
-      panfrost_bo_wait(rsrc->bo, INT64_MAX, false);
+      panfrost_resource_wait(rsrc, ctx, INT64_MAX, false);
       uint64_t *timestamp = (uint64_t *)rsrc->bo->ptr.cpu;
 
       vresult->u64 = pan_gpu_time_to_ns(dev, *timestamp);
@@ -853,7 +854,7 @@ panfrost_get_query_result(struct pipe_context *pipe, struct pipe_query *q,
 
    case PIPE_QUERY_TIME_ELAPSED: {
       panfrost_flush_writer(ctx, rsrc, "Time elapsed query");
-      panfrost_bo_wait(rsrc->bo, INT64_MAX, false);
+      panfrost_resource_wait(rsrc, ctx, INT64_MAX, false);
       uint64_t *timestamp = (uint64_t *)rsrc->bo->ptr.cpu;
 
       vresult->u64 = pan_gpu_time_to_ns(dev, timestamp[1] - timestamp[0]);
diff --git a/src/gallium/drivers/panfrost/pan_resource.c b/src/gallium/drivers/panfrost/pan_resource.c
index 4a402ef6be89..a70c741a5358 100644
--- a/src/gallium/drivers/panfrost/pan_resource.c
+++ b/src/gallium/drivers/panfrost/pan_resource.c
@@ -1371,6 +1371,14 @@ panfrost_load_tiled_images(struct panfrost_transfer *transfer,
    }
 }
 
+bool
+panfrost_resource_wait(struct panfrost_resource *rsrc,
+                       UNUSED struct panfrost_context *ctx, int64_t timeout_ns,
+                       bool wait_readers)
+{
+   return panfrost_bo_wait(rsrc->bo, timeout_ns, wait_readers);
+}
+
 #if MESA_DEBUG
 
 void
@@ -1408,7 +1416,7 @@ pan_dump_resource(struct panfrost_context *ctx, struct panfrost_resource *rsc)
    }
 
    panfrost_flush_writer(ctx, linear, "dump image");
-   panfrost_bo_wait(linear->bo, INT64_MAX, false);
+   panfrost_resource_wait(linear, ctx, INT64_MAX, false);
 
    if (!panfrost_bo_mmap(linear->bo)) {
       static unsigned frame_count = 0;
@@ -1544,7 +1552,7 @@ panfrost_ptr_map(struct pipe_context *pctx, struct pipe_resource *resource,
           (valid || panfrost_any_batch_writes_rsrc(ctx, rsrc))) {
          pan_blit_to_staging(pctx, transfer);
          panfrost_flush_writer(ctx, staging, "AFBC/AFRC tex read staging blit");
-         panfrost_bo_wait(staging->bo, INT64_MAX, false);
+         panfrost_resource_wait(staging, ctx, INT64_MAX, false);
       }
 
       if (panfrost_bo_mmap(staging->bo))
@@ -1591,7 +1599,7 @@ panfrost_ptr_map(struct pipe_context *pctx, struct pipe_resource *resource,
        */
 
       panfrost_flush_writer(ctx, rsrc, "Shadow resource creation");
-      panfrost_bo_wait(bo, INT64_MAX, false);
+      panfrost_resource_wait(rsrc, ctx, INT64_MAX, false);
 
       create_new_bo = true;
       copy_resource = !(usage & PIPE_MAP_DISCARD_WHOLE_RESOURCE);
@@ -1616,7 +1624,7 @@ panfrost_ptr_map(struct pipe_context *pctx, struct pipe_resource *resource,
        * batches), we try to allocate a new one to avoid waiting.
        */
       if (panfrost_any_batch_reads_rsrc(ctx, rsrc) ||
-          !panfrost_bo_wait(bo, 0, true)) {
+          !panfrost_resource_wait(rsrc, ctx, 0, true)) {
          /* We want the BO to be MMAPed. */
          uint32_t flags = bo->flags & ~PAN_BO_DELAY_MMAP;
          struct panfrost_bo *newbo = NULL;
@@ -1657,16 +1665,16 @@ panfrost_ptr_map(struct pipe_context *pctx, struct pipe_resource *resource,
              */
             panfrost_flush_batches_accessing_rsrc(
                ctx, rsrc, "Resource access with high memory pressure");
-            panfrost_bo_wait(bo, INT64_MAX, true);
+            panfrost_resource_wait(rsrc, ctx, INT64_MAX, true);
          }
       }
    } else if (!(usage & PIPE_MAP_UNSYNCHRONIZED)) {
       if (usage & PIPE_MAP_WRITE) {
          panfrost_flush_batches_accessing_rsrc(ctx, rsrc, "Synchronized write");
-         panfrost_bo_wait(bo, INT64_MAX, true);
+         panfrost_resource_wait(rsrc, ctx, INT64_MAX, true);
       } else if (usage & PIPE_MAP_READ) {
          panfrost_flush_writer(ctx, rsrc, "Synchronized read");
-         panfrost_bo_wait(bo, INT64_MAX, false);
+         panfrost_resource_wait(rsrc, ctx, INT64_MAX, false);
       }
    }
 
@@ -2193,7 +2201,7 @@ pan_resource_afbcp_update(struct panfrost_context *ctx,
 
    /* 1st async AFBC-P step: get payload sizes. */
    if (!prsrc->afbcp->layout_bo) {
-      if (!panfrost_bo_wait(prsrc->bo, 0, false))
+      if (!panfrost_resource_wait(prsrc, ctx, 0, false))
          return;
       if (!pan_resource_afbcp_get_payload_sizes(ctx, prsrc))
          goto stop_packing;
@@ -2218,7 +2226,7 @@ pan_resource_afbcp_update(struct panfrost_context *ctx,
 
    /* 3rd async AFBC-P step: pack. */
    if (!prsrc->afbcp->packed_bo) {
-      if (!panfrost_bo_wait(prsrc->bo, 0, false) ||
+      if (!panfrost_resource_wait(prsrc, ctx, 0, false) ||
           !panfrost_bo_wait(prsrc->afbcp->layout_bo, 0, false))
          return;
       if (!pan_resource_afbcp_pack(ctx, prsrc))
diff --git a/src/gallium/drivers/panfrost/pan_resource.h b/src/gallium/drivers/panfrost/pan_resource.h
index 047dc9fbf853..3f3f14cccd5d 100644
--- a/src/gallium/drivers/panfrost/pan_resource.h
+++ b/src/gallium/drivers/panfrost/pan_resource.h
@@ -217,4 +217,11 @@ void pan_resource_modifier_legalize(struct panfrost_context *ctx,
 void pan_dump_resource(struct panfrost_context *ctx,
                        struct panfrost_resource *rsc);
 
+void panfrost_blit_no_afbc_legalization(struct pipe_context *pipe,
+                                        const struct pipe_blit_info *info);
+
+bool panfrost_resource_wait(struct panfrost_resource *rsrc,
+                            struct panfrost_context *ctx, int64_t timeout_ns,
+                            bool wait_readers);
+
 #endif /* PAN_RESOURCE_H */
-- 
GitLab


From 0f5c7857731822576f8a144c424c62c26f41e54a Mon Sep 17 00:00:00 2001
From: Boris Brezillon <boris.brezillon@collabora.com>
Date: Thu, 23 Apr 2026 09:44:09 +0200
Subject: [PATCH 5/6] panfrost: Move BO access tracking to the context layer

With multi-context being able to concurrently access pipe_resource
objects, it becomes unsafe to have a panfrost_bo::gpu_access flag
that's encoding any access across all contexts. One way would be to
protect this ::gpu_access field and the wait logic that's using it
with a lock, but then it would serialize waits across contexts, even
though they might have different types (writers-only vs all) and
different timeouts.

The other option is to track accesses per-context, and turn
panfrost_bo_wait() into a pure KMD non-optimized wait, which is what
this commit is doing.

There's a few things to note here:

- the panfrost_bo_wait() that's done in the fetch_from_cache path is no
  longer optimized to avoid the IOCTL when the BO has no pending access.
  It's probably okay because we're doing a bunch of other IOCTLs in that
  path already, so if it was syscall bound, it would already be a problem
- we were previously flagging internal BOs with pending RW accesses in
  csf_attach_sync_points() and panfrost_pool_get_bo_handles(). With
  pan_bo now unconditionally calling pan_kmod_bo_wait(), this is no
  longer needed, because we know these per-batch mem pools will get
  recycled through the pan_bo cache.

Signed-off-by: Boris Brezillon <boris.brezillon@collabora.com>
Reviewed-by: Eric Guo <Eric.guo@nxp.com>
---
 src/gallium/drivers/panfrost/pan_bo.c       | 62 ++++++++-------------
 src/gallium/drivers/panfrost/pan_bo.h       | 32 ++++++-----
 src/gallium/drivers/panfrost/pan_context.c  |  1 +
 src/gallium/drivers/panfrost/pan_context.h  | 60 ++++++++++++++++++++
 src/gallium/drivers/panfrost/pan_csf.c      | 11 +---
 src/gallium/drivers/panfrost/pan_jm.c       | 12 +---
 src/gallium/drivers/panfrost/pan_job.c      |  4 +-
 src/gallium/drivers/panfrost/pan_job.h      | 12 ++++
 src/gallium/drivers/panfrost/pan_mempool.c  | 16 ++----
 src/gallium/drivers/panfrost/pan_resource.c | 55 +++++++++++++++++-
 10 files changed, 174 insertions(+), 91 deletions(-)

diff --git a/src/gallium/drivers/panfrost/pan_bo.c b/src/gallium/drivers/panfrost/pan_bo.c
index d0ce95cb66a3..8398bed65b2a 100644
--- a/src/gallium/drivers/panfrost/pan_bo.c
+++ b/src/gallium/drivers/panfrost/pan_bo.c
@@ -50,6 +50,24 @@ to_kmod_bo_flags(uint32_t flags)
    return kmod_bo_flags;
 }
 
+static void
+reset_bo_slot(struct panfrost_bo *bo)
+{
+   *bo = (struct panfrost_bo){
+      .seqno = bo->seqno,
+   };
+}
+
+UNUSED static bool
+bo_slot_is_free(struct panfrost_bo *bo)
+{
+   struct panfrost_bo expected = {
+      .seqno = bo->seqno,
+   };
+
+   return !memcmp(bo, &expected, sizeof(*bo));
+}
+
 static struct panfrost_bo *
 panfrost_bo_alloc(struct panfrost_device *dev, size_t size, uint32_t flags,
                   const char *label)
@@ -66,7 +84,7 @@ panfrost_bo_alloc(struct panfrost_device *dev, size_t size, uint32_t flags,
       goto err_alloc;
 
    bo = pan_lookup_bo(dev, kmod_bo->handle);
-   assert(!memcmp(bo, &((struct panfrost_bo){0}), sizeof(*bo)));
+   assert(bo_slot_is_free(bo));
    bo->kmod_bo = kmod_bo;
 
    struct pan_kmod_vm_op vm_op = {
@@ -96,8 +114,7 @@ panfrost_bo_alloc(struct panfrost_device *dev, size_t size, uint32_t flags,
 
 err_bind:
    pan_kmod_bo_put(kmod_bo);
-   /* BO will be freed with the sparse array, but zero to indicate free */
-   memset(bo, 0, sizeof(*bo));
+   reset_bo_slot(bo);
 err_alloc:
    return NULL;
 }
@@ -112,7 +129,7 @@ panfrost_bo_free(struct panfrost_bo *bo)
    uint64_t gpu_va = bo->ptr.gpu;
 
    /* BO will be freed with the sparse array, but zero to indicate free */
-   memset(bo, 0, sizeof(*bo));
+   reset_bo_slot(bo);
 
    struct pan_kmod_vm_op vm_op = {
       .type = PAN_KMOD_VM_OP_TYPE_UNMAP,
@@ -130,42 +147,6 @@ panfrost_bo_free(struct panfrost_bo *bo)
    pan_kmod_bo_put(kmod_bo);
 }
 
-/* Returns true if the BO is ready, false otherwise.
- * access_type is encoding the type of access one wants to ensure is done.
- * Waiting is always done for writers, but if wait_readers is set then readers
- * are also waited for.
- */
-bool
-panfrost_bo_wait(struct panfrost_bo *bo, int64_t timeout_ns, bool wait_readers)
-{
-   PAN_TRACE_FUNC(PAN_TRACE_GL_BO);
-
-   /* If the BO has been exported or imported we can't rely on the cached
-    * state, we need to call the WAIT_BO ioctl.
-    */
-   if (!(bo->flags & PAN_BO_SHARED)) {
-      /* If ->gpu_access is 0, the BO is idle, no need to wait. */
-      if (!bo->gpu_access)
-         return true;
-
-      /* If the caller only wants to wait for writers and no
-       * writes are pending, we don't have to wait.
-       */
-      if (!wait_readers && !(bo->gpu_access & PAN_BO_ACCESS_WRITE))
-         return true;
-   }
-
-   if (pan_kmod_bo_wait(bo->kmod_bo, timeout_ns, !wait_readers)) {
-      /* Set gpu_access to 0 so that the next call to bo_wait()
-       * doesn't have to call the WAIT_BO ioctl.
-       */
-      bo->gpu_access = 0;
-      return true;
-   }
-
-   return false;
-}
-
 /* Helper to calculate the bucket index of a BO */
 
 static unsigned
@@ -420,6 +401,7 @@ panfrost_bo_create(struct panfrost_device *dev, size_t size, uint32_t flags,
    }
 
    panfrost_bo_set_label(bo, label);
+   bo->seqno++;
 
    return bo;
 }
diff --git a/src/gallium/drivers/panfrost/pan_bo.h b/src/gallium/drivers/panfrost/pan_bo.h
index b3a5d9f58983..2c5f627904eb 100644
--- a/src/gallium/drivers/panfrost/pan_bo.h
+++ b/src/gallium/drivers/panfrost/pan_bo.h
@@ -38,15 +38,6 @@
  * PAN_BO_SHARED if the BO has not been exported yet */
 #define PAN_BO_SHAREABLE (1 << 5)
 
-/* GPU access flags */
-
-/* BO is being read/written by the GPU */
-#define PAN_BO_ACCESS_READ  (1 << 0)
-#define PAN_BO_ACCESS_WRITE (1 << 1)
-#define PAN_BO_ACCESS_RW    (PAN_BO_ACCESS_READ | PAN_BO_ACCESS_WRITE)
-
-typedef uint8_t pan_bo_access;
-
 struct panfrost_device;
 
 struct panfrost_bo {
@@ -74,11 +65,11 @@ struct panfrost_bo {
 
    uint32_t flags;
 
-   /* Combination of PAN_BO_ACCESS_{READ,WRITE} flags encoding pending
-    * GPU accesses to this BO. Useful to avoid calling the WAIT_BO ioctl
-    * when the BO is idle.
+   /* Sequence number that gets incremented every time a BO is allocated.
+    * Used to identify stale data in the per-BO access tracking done at the
+    * panfrost_context level.
     */
-   uint32_t gpu_access;
+   uint32_t seqno;
 
    /* Driver-provided human-readable description of the BO for debugging.
     * Consists of a BO name and optional additional information, like
@@ -102,8 +93,19 @@ panfrost_bo_handle(struct panfrost_bo *bo)
 struct panfrost_bo *panfrost_bo_from_kmod_bo(struct panfrost_device *dev,
                                              struct pan_kmod_bo *kmod_bo);
 
-bool panfrost_bo_wait(struct panfrost_bo *bo, int64_t timeout_ns,
-                      bool wait_readers);
+/* Returns true if the BO is ready, false otherwise.
+ * access_type is encoding the type of access one wants to ensure is done.
+ * Waiting is always done for writers, but if wait_readers is set then readers
+ * are also waited for.
+ */
+static inline bool
+panfrost_bo_wait(struct panfrost_bo *bo, int64_t timeout_ns, bool wait_readers)
+{
+   PAN_TRACE_FUNC(PAN_TRACE_GL_BO);
+
+   return pan_kmod_bo_wait(bo->kmod_bo, timeout_ns, !wait_readers);
+}
+
 void panfrost_bo_reference(struct panfrost_bo *bo);
 void panfrost_bo_unreference(struct panfrost_bo *bo);
 struct panfrost_bo *panfrost_bo_create(struct panfrost_device *dev, size_t size,
diff --git a/src/gallium/drivers/panfrost/pan_context.c b/src/gallium/drivers/panfrost/pan_context.c
index b48c9720947f..4dbc7db68c8e 100644
--- a/src/gallium/drivers/panfrost/pan_context.c
+++ b/src/gallium/drivers/panfrost/pan_context.c
@@ -1176,6 +1176,7 @@ panfrost_create_context(struct pipe_screen *screen, void *priv, unsigned flags)
    ctx->sample_mask = ~0;
    ctx->active_queries = true;
 
+   util_dynarray_init(&ctx->bo_access, ctx);
    util_dynarray_init(&ctx->global_buffers, ctx);
 
    /* Sync object/FD used for NATIVE_FENCE_FD. */
diff --git a/src/gallium/drivers/panfrost/pan_context.h b/src/gallium/drivers/panfrost/pan_context.h
index 133c84ad8b8e..00f16d19c029 100644
--- a/src/gallium/drivers/panfrost/pan_context.h
+++ b/src/gallium/drivers/panfrost/pan_context.h
@@ -25,6 +25,7 @@
 #include "util/format/u_formats.h"
 #include "util/hash_table.h"
 #include "util/simple_mtx.h"
+#include "util/u_dynarray.h"
 #include "util/u_blitter.h"
 #include "util/u_printf.h"
 
@@ -252,6 +253,16 @@ struct panfrost_context {
    struct panfrost_perfetto_state perfetto;
 #endif
    uint32_t submit_count; /* monotonic submit ID for perfetto */
+
+   /* Used to track in-flight BO accesses done by this context (job
+    * accessing those BOs have been submitted, but we don't know if
+    * they completed yet).
+    * The array is indexed by BO handles and is never shrunk.
+    * Each uint32_t entry encodes the pending access type (R/W) and the
+    * BO seqno at the time of this access. The seqno allows us to
+    * detect stale entries when a BO or its handle is recycled.
+    */
+   struct util_dynarray bo_access;
 };
 
 /* Corresponds to the CSO */
@@ -614,4 +625,53 @@ void panfrost_track_image_access(struct panfrost_batch *batch,
 
 void panfrost_context_reinit(struct panfrost_context *ctx);
 
+static inline void
+panfrost_context_report_bo_access(struct panfrost_context *ctx,
+                                  struct panfrost_bo *bo, uint32_t access)
+{
+   /* Skip if per-context tracking is not requested or if the BO is shared. */
+   if (!(access & PAN_BO_ACCESS_PER_CTX_TRACKING || (bo->flags & PAN_BO_SHARED)))
+      return;
+
+   ASSERTED void *new_bo_access;
+   uint32_t handle = panfrost_bo_handle(bo);
+
+   new_bo_access = util_dynarray_resize_zero(&ctx->bo_access, uint32_t, handle + 1);
+   if (!new_bo_access)
+      return;
+
+   uint32_t *access_ptr = util_dynarray_element(&ctx->bo_access, uint32_t, handle);
+   uint32_t seqno = *access_ptr >> 2;
+   uint32_t pending = *access_ptr & BITFIELD_MASK(2);
+
+   static_assert(PAN_BO_ACCESS_RW == BITFIELD_MASK(2),
+                 "PAN_BO_ACCESS_RW expected to cover the lower 2-bits");
+
+   access &= PAN_BO_ACCESS_RW;
+
+   /* We don't have an easy way to know when the BO leaves the context,
+    * nor do we have a way to get back to contexts that might have been
+    * using the BO when the resource is destroyed. This means we're left
+    * with potential stale data in the bo_access array, which we check
+    * validity of based on a per-BO seqno that gets incremented every
+    * time a BO is allocated (handle re-allocated, or BO taken from the
+    * BO cache, which forces a wait-on-everything). If that happens,
+    * the pending accesses are reset, and the seqno gets updated to match
+    * the BO seqno.
+    */
+   if (seqno != bo->seqno)
+      seqno = bo->seqno;
+   else
+      access |= pending;
+
+   /* We lose the 2 MSB of the seqno here, but 30 bits should be more
+    * than enough to prevent spurious wrap around messing up with the
+    * stale entry detection. Worst case scenario, we really have 2^30
+    * rotations on this handle, and we wrongly consider the entry valid
+    * and wait when we could have skipped the wait => that's not the end
+    * of the world.
+    */
+   *access_ptr = (seqno << 2) | access;
+}
+
 #endif
diff --git a/src/gallium/drivers/panfrost/pan_csf.c b/src/gallium/drivers/panfrost/pan_csf.c
index 4b7607942b50..9dbb906e87bd 100644
--- a/src/gallium/drivers/panfrost/pan_csf.c
+++ b/src/gallium/drivers/panfrost/pan_csf.c
@@ -627,13 +627,6 @@ csf_submit_collect_wait_ops(struct panfrost_batch *batch,
       if (!flags)
          continue;
 
-      /* Update the BO access flags so that panfrost_bo_wait() knows
-       * about all pending accesses.
-       * We only keep the READ/WRITE info since this is all the BO
-       * wait logic cares about.
-       * We also preserve existing flags as this batch might not
-       * be the first one to access the BO.
-       */
       struct panfrost_bo *bo = pan_lookup_bo(dev, i);
 
       ret = panthor_kmod_bo_get_sync_point(bo->kmod_bo, &bo_sync_handle,
@@ -708,7 +701,6 @@ csf_attach_sync_points(struct panfrost_batch *batch, uint32_t vm_sync_handle,
     * be written by the GPU to keep things simple.
     */
    util_dynarray_foreach(&batch->pool.bos, struct panfrost_bo *, bo) {
-      (*bo)->gpu_access |= PAN_BO_ACCESS_RW;
       ret = panthor_kmod_bo_attach_sync_point((*bo)->kmod_bo, vm_sync_handle,
                                               vm_sync_signal_point, true);
       if (ret)
@@ -717,7 +709,6 @@ csf_attach_sync_points(struct panfrost_batch *batch, uint32_t vm_sync_handle,
 
    util_dynarray_foreach(&batch->csf.cs_chunk_pool.bos, struct panfrost_bo *,
                          bo) {
-      (*bo)->gpu_access |= PAN_BO_ACCESS_RW;
       ret = panthor_kmod_bo_attach_sync_point((*bo)->kmod_bo, vm_sync_handle,
                                               vm_sync_signal_point, true);
       if (ret)
@@ -734,7 +725,7 @@ csf_attach_sync_points(struct panfrost_batch *batch, uint32_t vm_sync_handle,
 
       struct panfrost_bo *bo = pan_lookup_bo(dev, i);
 
-      bo->gpu_access |= flags & (PAN_BO_ACCESS_RW);
+      panfrost_context_report_bo_access(ctx, bo, flags);
       ret = panthor_kmod_bo_attach_sync_point(bo->kmod_bo, vm_sync_handle,
                                               vm_sync_signal_point,
                                               flags & PAN_BO_ACCESS_WRITE);
diff --git a/src/gallium/drivers/panfrost/pan_jm.c b/src/gallium/drivers/panfrost/pan_jm.c
index 6d2c0453f5c3..d14fc05082fb 100644
--- a/src/gallium/drivers/panfrost/pan_jm.c
+++ b/src/gallium/drivers/panfrost/pan_jm.c
@@ -117,17 +117,7 @@ jm_submit_jc(struct panfrost_batch *batch, uint64_t first_job_desc,
 
       assert(submit.bo_handle_count < batch->num_bos);
       bo_handles[submit.bo_handle_count++] = i;
-
-      /* Update the BO access flags so that panfrost_bo_wait() knows
-       * about all pending accesses.
-       * We only keep the READ/WRITE info since this is all the BO
-       * wait logic cares about.
-       * We also preserve existing flags as this batch might not
-       * be the first one to access the BO.
-       */
-      struct panfrost_bo *bo = pan_lookup_bo(dev, i);
-
-      bo->gpu_access |= flags[i] & (PAN_BO_ACCESS_RW);
+      panfrost_context_report_bo_access(ctx, pan_lookup_bo(dev, i), flags[i]);
    }
 
    panfrost_pool_get_bo_handles(&batch->pool,
diff --git a/src/gallium/drivers/panfrost/pan_job.c b/src/gallium/drivers/panfrost/pan_job.c
index a3da7ddd86ad..51e4bfae53f0 100644
--- a/src/gallium/drivers/panfrost/pan_job.c
+++ b/src/gallium/drivers/panfrost/pan_job.c
@@ -358,7 +358,7 @@ panfrost_batch_read_rsrc(struct panfrost_batch *batch,
    if (rsrc->base.next)
       panfrost_batch_read_rsrc(batch, pan_resource(rsrc->base.next));
 
-   uint32_t access = PAN_BO_ACCESS_READ;
+   uint32_t access = PAN_BO_ACCESS_READ | PAN_BO_ACCESS_PER_CTX_TRACKING;
 
    pan_resource_update_access(batch->ctx, rsrc, false);
 
@@ -382,7 +382,7 @@ panfrost_batch_write_rsrc(struct panfrost_batch *batch,
    if (rsrc->base.next)
       panfrost_batch_write_rsrc(batch, pan_resource(rsrc->base.next));
 
-   uint32_t access = PAN_BO_ACCESS_WRITE;
+   uint32_t access = PAN_BO_ACCESS_WRITE | PAN_BO_ACCESS_PER_CTX_TRACKING;
 
    pan_resource_update_access(batch->ctx, rsrc, true);
 
diff --git a/src/gallium/drivers/panfrost/pan_job.h b/src/gallium/drivers/panfrost/pan_job.h
index 81e47cf36a4d..ea8a9e250702 100644
--- a/src/gallium/drivers/panfrost/pan_job.h
+++ b/src/gallium/drivers/panfrost/pan_job.h
@@ -17,6 +17,18 @@
 #include "pan_resource.h"
 #include "util/perf/u_trace.h"
 
+/* BO is being read/written by the GPU */
+#define PAN_BO_ACCESS_READ  BITFIELD_BIT(0)
+#define PAN_BO_ACCESS_WRITE BITFIELD_BIT(1)
+#define PAN_BO_ACCESS_RW    (PAN_BO_ACCESS_READ | PAN_BO_ACCESS_WRITE)
+
+/* Set when the BO access needs to be recorded to the
+ * panfrost_context::bo_access array at submission time.
+ */
+#define PAN_BO_ACCESS_PER_CTX_TRACKING BITFIELD_BIT(2)
+
+typedef uint8_t pan_bo_access;
+
 /* A panfrost_batch corresponds to a bound FBO we're rendering to,
  * collecting over multiple draws. */
 
diff --git a/src/gallium/drivers/panfrost/pan_mempool.c b/src/gallium/drivers/panfrost/pan_mempool.c
index 1c902616a226..fbdd1bb2e943 100644
--- a/src/gallium/drivers/panfrost/pan_mempool.c
+++ b/src/gallium/drivers/panfrost/pan_mempool.c
@@ -7,6 +7,7 @@
 #include <unistd.h>
 #include <sys/mman.h>
 
+#include "pan_context.h"
 #include "pan_device.h"
 #include "pan_mempool.h"
 #include "pan_trace.h"
@@ -102,17 +103,10 @@ panfrost_pool_get_bo_handles(struct panfrost_pool *pool, uint32_t *handles)
 
    unsigned idx = 0;
    util_dynarray_foreach(&pool->bos, struct panfrost_bo *, bo) {
-      assert(panfrost_bo_handle(*bo) > 0);
-      handles[idx++] = panfrost_bo_handle(*bo);
-
-      /* Update the BO access flags so that panfrost_bo_wait() knows
-       * about all pending accesses.
-       * We only keep the READ/WRITE info since this is all the BO
-       * wait logic cares about.
-       * We also preserve existing flags as this batch might not
-       * be the first one to access the BO.
-       */
-      (*bo)->gpu_access |= PAN_BO_ACCESS_RW;
+      uint32_t handle = panfrost_bo_handle(*bo);
+
+      assert(handle > 0);
+      handles[idx++] = handle;
    }
 }
 
diff --git a/src/gallium/drivers/panfrost/pan_resource.c b/src/gallium/drivers/panfrost/pan_resource.c
index a70c741a5358..7cf1f86933b0 100644
--- a/src/gallium/drivers/panfrost/pan_resource.c
+++ b/src/gallium/drivers/panfrost/pan_resource.c
@@ -1371,12 +1371,63 @@ panfrost_load_tiled_images(struct panfrost_transfer *transfer,
    }
 }
 
+/* panfrost_resource_wait() only waits on panfrost_resource::bo even though
+ * there might be multiple other BOs attached to the resource.
+ *
+ * Here are some aspects that are worth noting:
+ * - rsrc::{separate_stencil,shadow_image}::bo is implicitly waited on because
+ *   panfrost_batch_{write,read}_rsrc() always record access to these resources
+ *   if they are present, so waiting on one of them is equivalent to waiting
+ *   on all of them
+ * - multiplanar resources should be covered because
+ *   panfrost_batch_{write,read}_rsrc() always record accesses on all planes
+ * - panfrost_resource::afbc::{layout,packed}_bo are not covered. They must
+ *   be waited on manually with a panfrost_bo_wait() call, because we don't
+ *   even track accesses to those a the panfrost_context::bo_access level
+ */
 bool
 panfrost_resource_wait(struct panfrost_resource *rsrc,
-                       UNUSED struct panfrost_context *ctx, int64_t timeout_ns,
+                       struct panfrost_context *ctx, int64_t timeout_ns,
                        bool wait_readers)
 {
-   return panfrost_bo_wait(rsrc->bo, timeout_ns, wait_readers);
+   /* The BO is shared, we have to do a KMD wait. */
+   if (rsrc->bo->flags & PAN_BO_SHARED)
+      return panfrost_bo_wait(rsrc->bo, timeout_ns, wait_readers);
+
+   uint32_t handle = panfrost_bo_handle(rsrc->bo);
+   uint32_t *pending_access = NULL;
+   uint32_t seqno = 0, access = 0;
+
+   if (util_dynarray_num_elements(&ctx->bo_access, uint32_t) > handle) {
+      pending_access = util_dynarray_element(&ctx->bo_access, uint32_t, handle);
+      seqno = *pending_access >> 2;
+      access = *pending_access & PAN_BO_ACCESS_RW;
+   }
+
+   uint32_t mask = wait_readers ? PAN_BO_ACCESS_RW : PAN_BO_ACCESS_WRITE;
+   bool ready = false;
+
+   /* If the seqno is zero, it means the entry was uninitialized, and we can't
+    * trust it because it might come from an allocation failure when we try to
+    * resize the bo_access array in panfrost_context_report_bo_access(). In
+    * that case, we just take the hit and do a KMD wait.
+    */
+   if (seqno > 0) {
+      /* If the seqno don't match, the BO has been returned and a fresh one
+       * allocated with the same handle. In that case, pan_bo guarantees the
+       * buffer is idle.
+       */
+      if (rsrc->bo->seqno != seqno || !(access & mask))
+         ready = true;
+   }
+
+   if (!ready)
+      ready = panfrost_bo_wait(rsrc->bo, timeout_ns, wait_readers);
+
+   if (ready && pending_access)
+      *pending_access &= ~mask;
+
+   return ready;
 }
 
 #if MESA_DEBUG
-- 
GitLab


From 65525f07a212fa178b0ed810f3e972006c6072ef Mon Sep 17 00:00:00 2001
From: Boris Brezillon <boris.brezillon@collabora.com>
Date: Thu, 23 Apr 2026 12:12:33 +0200
Subject: [PATCH 6/6] panfrost: Make BO label/flags update threadsafe

pipe_resources and the BO backing them can be shared across contexts,
so we need to make sure the panfrost_bo updates happening after
allocation are threadsafe, which covers the flags update happening in
panfrost_bo_export() and the label update in panfrost_bo_replace_label().

Signed-off-by: Boris Brezillon <boris.brezillon@collabora.com>
Reviewed-by: Eric Guo <Eric.guo@nxp.com>
---
 src/gallium/drivers/panfrost/pan_bo.c | 14 +++++++++++++-
 src/gallium/drivers/panfrost/pan_bo.h |  3 +++
 2 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/src/gallium/drivers/panfrost/pan_bo.c b/src/gallium/drivers/panfrost/pan_bo.c
index 8398bed65b2a..995224a7924f 100644
--- a/src/gallium/drivers/panfrost/pan_bo.c
+++ b/src/gallium/drivers/panfrost/pan_bo.c
@@ -110,6 +110,7 @@ panfrost_bo_alloc(struct panfrost_device *dev, size_t size, uint32_t flags,
    bo->ptr.gpu = vm_op.va.start;
    bo->flags = flags;
    bo->dev = dev;
+   simple_mtx_init(&bo->lock, mtx_plain);
    return bo;
 
 err_bind:
@@ -128,6 +129,11 @@ panfrost_bo_free(struct panfrost_bo *bo)
    struct pan_kmod_vm *vm = bo->dev->kmod.vm;
    uint64_t gpu_va = bo->ptr.gpu;
 
+   /* reset_bo_slot() will undo the poisoning done in simple_mtx_destroy(), but
+    * we need it for the valgrind annotation.
+    */
+   simple_mtx_destroy(&bo->lock);
+
    /* BO will be freed with the sparse array, but zero to indicate free */
    reset_bo_slot(bo);
 
@@ -533,8 +539,11 @@ panfrost_bo_export(struct panfrost_bo *bo)
    PAN_TRACE_FUNC(PAN_TRACE_GL_BO);
 
    int ret = pan_kmod_bo_export(bo->kmod_bo);
-   if (ret >= 0)
+   if (ret >= 0) {
+      simple_mtx_lock(&bo->lock);
       bo->flags |= PAN_BO_SHARED;
+      simple_mtx_unlock(&bo->lock);
+   }
 
    return ret;
 }
@@ -556,6 +565,7 @@ const char *
 panfrost_bo_replace_label(struct panfrost_bo *bo, const char *label,
                           bool set_kernel_label)
 {
+   simple_mtx_lock(&bo->lock);
    const char *old_label = bo->label;
 
    bo->label = label;
@@ -563,5 +573,7 @@ panfrost_bo_replace_label(struct panfrost_bo *bo, const char *label,
    if (set_kernel_label)
       pan_kmod_set_bo_label(bo->dev->kmod.dev, bo->kmod_bo, label);
 
+   simple_mtx_unlock(&bo->lock);
+
    return old_label;
 }
diff --git a/src/gallium/drivers/panfrost/pan_bo.h b/src/gallium/drivers/panfrost/pan_bo.h
index 2c5f627904eb..fedeb7c6c1fa 100644
--- a/src/gallium/drivers/panfrost/pan_bo.h
+++ b/src/gallium/drivers/panfrost/pan_bo.h
@@ -76,6 +76,9 @@ struct panfrost_bo {
     * pixel format or buffer modifiers.
     */
    const char *label;
+
+   /* Lock used to protect concurrent access to flags and label. */
+   simple_mtx_t lock;
 };
 
 static inline size_t
-- 
GitLab

